File: clang/ads/linked_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
  int data;
  struct Node* next;
} Node;

Node* create_node(int data);
void  insert_front(Node** head, int data);
void  insert_end(Node** head, int data);
void  insert_at(Node** head, int data, int position);
void  delete_first(Node** head);
void  delete_last(Node** head);
void  delete_at(Node** head, int position);
void  print(Node* head);

Node* create_node(int data){
  Node* node = malloc(sizeof(Node));
  node->data = data;
  node->next = NULL;
  return node;
}

void insert_front(Node** head, int data){
  Node* node = create_node(data);
  node->next = *head;
  *head = node;
}

void insert_end(Node** head, int data){
  Node* node = create_node(data);
  if (*head == NULL){
    *head = node;
    return;
  }

  Node* temp = *head;
  while (temp->next != NULL) {
    temp = temp->next;
  }
  
  temp->next = node;
}

void insert_at(Node** head, int data, int position){
  Node* node = create_node(data);
  if (position == 0){
    insert_front(head, data);
    return;
  }

  Node* temp = *head;
  for (int i = 0; temp != NULL && i < position -1; i++) {
    temp = temp->next;
  }

  if (temp == NULL){
    printf("Position out of range\n");
    free(node);
    return;
  }

  node->next = temp->next;
  temp->next = node;
}

void delete_first(Node** head){
  if (*head == NULL){
    printf("List is empty\n");
    return;
  }

  Node* temp = *head;
  *head = temp->next;
  free(temp);
}

void delete_last(Node** head){
  if (*head == NULL){
    printf("List is empty\n");
    return;
  }

  Node* temp = *head;
  if (temp->next == NULL){
    free(temp);
    *head = NULL;
    return;
  }

  while (temp->next->next != NULL) {
    temp = temp->next;
  }

  free(temp->next);
  temp->next = NULL;
}

void delete_at(Node** head, int position){
  if (*head == NULL){
    printf("List is empty\n");
    return;
  }

  Node* temp = *head;
  if (position == 0){
    delete_first(head);
    return;
  }

  for (int i = 0; temp != NULL && i < position -1; i++) {
    temp = temp->next;
  }

  if (temp == NULL || temp->next == NULL){
    printf("Position out of range\n");
    return;
  }

  Node* next = temp->next->next;
  free(temp->next);
  temp->next = next;
}

void print(Node* head){
  Node* temp = head;
  while (temp != NULL) {
    printf("%d -> ", temp->data);
    temp = temp->next;
  }
  printf("NULL\n");
}

File: nim/ads/linked_list.nim
import strformat

type Node = ref object
  data: int
  next: Node

func newNode(data: int): Node =
  Node(data: data)

proc insertFront(head: var Node, data: int) =
  let node = newNode(data)
  node.next = head
  head = node

proc insertEnd(head: var Node, data: int) =
  let node = newNode(data)
  if head.isNil():
    head = node
    return

  var temp = head
  while not temp.next.isNil():
    temp = temp.next

  temp.next = node

proc inserAt(head: var Node, data: int, position: int) =
  let node = newNode(data)
  if position == 0: insertFront(head, data)

  var temp = head
  for i in 0 ..< position: temp = temp.next

  if temp.isNil():
    echo "Position out of range"
    return

  node.next = temp.next
  temp.next = node

proc deleteFirst(head: var Node) =
  if head.isNil():
    echo "List is empty"
    return

  let temp = head
  head = temp.next

proc deleteLast(head: var Node) =
  if head.isNil():
    echo "List is empty"
    return

  var temp = head
  if temp.next.isNil():
    head = nil
    return

  while temp.next.next != nil:
    temp = temp.next

  temp.next = nil

proc deleteAt(head: var Node, position: int) = 
  if head.isNil():
    echo "List is empty"
    return

  var temp = head
  if position == 0:
    deleteFirst(head)
    return

  for i in 0 ..< position:
    temp = temp.next

  let next = temp.next.next
  temp.next = next

proc print(head: var Node) =
  var temp = head
  stdout.write "|"
  while temp != nil:
    stdout.write &" {temp.data} |"
    temp = temp.next
  echo()

proc main() =
  var node = newNode(0)
  node.print()

  node.insertFront(1)
  node.print()

  node.insertEnd(2)
  node.print()

  node.inserAt(3, 1)

  echo()
  node.print()
  echo()

  node.deleteAt(1)
  node.print()

  node.deleteLast()
  node.print()

  node.deleteFirst()
  node.print()

when isMainModule:
  main()
